specification/dartLangSpec.tex

```diff
@@ -15646,17 +15648,24 @@ \subsubsection{Yield}
 \end{itemize}
 
 \LMHash{}%
-It is a compile-time error if a yield statement appears in a function that is not a generator function.
+It is a compile-time error if a yield statement appears in
+a function that is not a generator function.
 
 \LMHash{}%
-Let $T$ be the static type of $e$ and let $f$ be the immediately enclosing function.
-It is a compile-time error if either:
-\begin{itemize}
-\item
-the body of $f$ is marked \code{\ASYNC*} and the type \code{Stream<T>} may not be assigned to the declared return type of $f$.
-\item
-the body of $f$ is marked \code{\SYNC*} and the type \code{Iterable<T>} may not be assigned to the declared return type of $f$.
-\end{itemize}
+Let $T_e$ be the static type of $e$ and
+let $f$ be the immediately enclosing function.
+\commentary{Let $T_f$ specify the type of elements that $f$ yields:}
+If the body of $f$ is marked \code{\ASYNC*} and
+the declared return type of $f$ implements \code{Stream<$U$>} for some $U$
+(\ref{interfaceSuperinterfaces})
+then let $T_f$ be $U$.
+Otherwise, if the body of $f$ is marked \code{\SYNC*} and
+the declared return type of $f$ implements \code{Iterable<$U$>} for some $U$
+then let $T_f$ be $U$.
+% $f$ is a generator function, so the return type must be a supertype of
+% `Iterable<U>`/`Stream<U>` for some `U`, so here it can only be a top type.
+Otherwise let $T_f$ be \DYNAMIC.
+It is a compile-time error if $T_e$ may not be assigned to $T_f$.
 
 
 \subsubsection{Yield-Each}
```

Comment:
Should we add an "it is a dynamic type error if the runtime type of $x$ is not a subtype of (the element type)" to the semantics of the yield*?

Reply:
Done.